### PR TITLE
[te] Fix casting of unsigned char, and abs(int)

### PR DESCRIPTION
**Pull request by bertmaher**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44152 Fix frac in CUDA fuser
* #44160 Prevent the TE fuser from getting datatypes it can't handle
* **#44157 [te] Fix casting of unsigned char, and abs(int)**

Differential Revision: [D23528507](https://our.internmc.facebook.com/intern/diff/D23528507)

### CODE REVIEW

File: test/test_jit_fuser_te.py [hunk 1]
```diff
@@ -1152,20 +1152,48 @@ def rand(dtype, device="cuda"):
             torch.int16,
             torch.int32,
             torch.int64,
-            torch.float16,
-            torch.bfloat16,
+            # torch.float16,
```
Comment (ZolotukhinM) [inline test/test_jit_fuser_te.py:1155]: Nit: maybe add a comment whether we want these to be supported at all?

File: torch/csrc/jit/tensorexpr/cuda_codegen.cpp [hunk 2]
```diff
@@ -186,7 +186,7 @@ void CudaPrinter::visit(const For* v) {
 }
 
 void CudaPrinter::visit(const Cast* v) {
-  os() << cudaDtypeCppString(v->dtype());
+  os() << "(" << cudaDtypeCppString(v->dtype()) << ")";
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/cuda_codegen.cpp:189]: Ouch, how did it work? :)

**Review by ZolotukhinM — APPROVED** [review]
Looks good! I guess you'll need to replace `self.assertEqual(ref, t(x))` with `np.testing.assert_allclose(...)` to appease the CI gods though.

File: test/test_jit_fuser_te.py (hunk 1 above)
Comment (bertmaher) [reply test/test_jit_fuser_te.py:1155]: Oh, yeah, float16 should be coming in soon-ish.  The commented lines are a transient state for this stack :-)

File: torch/csrc/jit/tensorexpr/cuda_codegen.cpp (hunk 2 above)
Comment (bertmaher) [reply torch/csrc/jit/tensorexpr/cuda_codegen.cpp:189]: I don't know when it became a thing but apparently you can cast by saying `float(x)`, like it's a constructor.  But that sorta doesn't parse if you do `unsigned char(x)`.  Clearly we need whitespace in identifiers! (I hope this reddit post is a joke: https://www.reddit.com/r/ProgrammerHumor/comments/4on81i/til_c_allows_u200b_zero_width_space_in_identifiers/ )

File: torch/csrc/jit/tensorexpr/cuda_codegen.cpp [hunk 3]
```diff
@@ -209,6 +209,9 @@ void CudaPrinter::visit(const Intrinsics* v) {
   if (returnType == ScalarType::Half || returnType == ScalarType::Float) {
     func_name = func_name + "f";
   }
+  if (v->op_type() == IntrinsicsOp::kFabs && is_integral(returnType)) {
```
Comment (ngimel) [inline torch/csrc/jit/tensorexpr/cuda_codegen.cpp:212]: does nvrtc not handle std::abs or ::abs? All this exp/expf is so last decade (and is not used in eager part of the codebase at all).

Comment (bertmaher) [reply torch/csrc/jit/tensorexpr/cuda_codegen.cpp:212]: I think it does, actually.  I don't know why we went the `+ "f"` route, except that maybe the old fuser uses it?  I'll take a note to use `std::` in a follow-up.

**Pull request closed without merging**